src/emailcontactstorage.cpp

```diff
@@ -0,0 +1,144 @@
+#include <algorithm>
+#include <QMailStore>
+#include "emailcontactstorage.h"
+
+Q_GLOBAL_STATIC(EmailContactStorage, storage)
+EmailContactStorage* EmailContactStorage::instance()
+{
+    if (storage != nullptr && storage->m_isInit == false)
+        storage->initialize();
+    return storage;
+}
+
+
+void EmailContactStorage::initialize()
+{
+    QMailStore *mailStore = QMailStore::instance();
+    Q_ASSERT(mailStore);
+
+    if (QMailStore::initializationState() != QMailStore::Initialized)
+        return;
+
+    QMailMessageKey incomingKey(QMailMessageKey::status(QMailMessage::Incoming));
+    QMailMessageIdList incomingIds = mailStore->queryMessages(incomingKey);
+    for (const auto& id: incomingIds) {
+        const QMailMessageMetaData metaData = mailStore->messageMetaData(id);
+        if (metaData.messageType() != QMailMessageMetaDataFwd::MessageType::Email)
+            continue;
+
+        const QMailTimeStamp timeStamp = metaData.date();
+        const QMailAddress from = metaData.from();
+        add(from, timeStamp);
+    }
+
+    QMailMessageKey outgoingKey(QMailMessageKey::status(QMailMessage::Outgoing));
+    QMailMessageIdList outgoingIds = mailStore->queryMessages(outgoingKey);
+    for (const auto& id: outgoingIds) {
+        const QMailMessageMetaData metaData = mailStore->messageMetaData(id);
+        if (metaData.messageType() != QMailMessageMetaDataFwd::MessageType::Email)
+            continue;
+
+        const QMailTimeStamp timeStamp = metaData.date();
+        const QList<QMailAddress> recipients = metaData.recipients();
+        for (const auto& address : recipients)
+            add(address, timeStamp);
+    }
+}
+
+
+int EmailContactStorage::count() const
+{
+    return m_storage.size();
+}
+
+
+QString EmailContactStorage::name(const int index) const
```

Comment:
Minor detail, but const on by-value parameters is a bit unnecessary noise. Shouldn't matter to the caller if the method internally adjusts its own copy.

Reply:
I like it when the parameters are constant.
1. It is symmetrical with constant references.
2.very rare, but helps inside the function